### PR TITLE
[CH-223]Support `TakeOrderedAndProjectExec`

### PR DESCRIPTION
**Pull request by lgbo-ustc**
## What changes were proposed in this pull request?

Support running `TakeOrderedAndProjectExec` in native engine.

## How was this patch tested?
 unit tests

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Comment by lgbo-ustc** [comment]
The final plan looks like
image

File: gluten-core/src/main/scala/io/glutenproject/execution/TakeOrderedAndProjectExecTransform.scala [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import org.apache.spark.sql.catalyst.plans.physical.{Partitioning, SinglePartition}
+import org.apache.spark.sql.catalyst.util.truncatedString
+import org.apache.spark.sql.execution.exchange.ShuffleExchangeExec
+import org.apache.spark.sql.execution.{ColumnarCollapseCodegenStages, ColumnarInputAdapter,
+  SparkPlan, UnaryExecNode}
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+import java.util.concurrent.atomic.AtomicInteger
+
+case class TakeOrderedAndProjectExecTransform (
```
Comment (rui-mo) [inline gluten-core/src/main/scala/io/glutenproject/execution/TakeOrderedAndProjectExecTransform.scala:32]: Maybe a rename to `TakeOrderedAndProjectExecTransformer` would make it better aligned with other operators.

File: backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCHSuite.scala [hunk 2]
```diff
@@ -273,6 +271,21 @@ class GlutenClickHouseTPCHSuite extends GlutenClickHouseTPCHAbstractSuite {
     TestUtils.compareAnswers(result, expected)
   }
 
+  test("test 'order by limit'") {
```
Comment (rui-mo) [inline backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCHSuite.scala:274]: Would you like to make this test a common test for both CH and Velox backend?

File: gluten-core/src/main/scala/io/glutenproject/execution/TakeOrderedAndProjectExecTransform.scala [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+
+  protected override def doExecuteColumnar(): RDD[ColumnarBatch] = {
+    val childRDD = child.executeColumnar()
+    val childRDDPartsNum = childRDD.getNumPartitions
+    if (childRDDPartsNum == 0) {
+      // ParallelCollectionRDD is private in spark, cannot be used here.
+      // But this is not a normal usage.
+      throw new UnsupportedOperationException(s"This operator doesn't support empty partitions")
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/execution/TakeOrderedAndProjectExecTransform.scala:69]: try to use `sparkContext.parallelize(Seq.empty, 1)`

File: gluten-core/src/main/scala/io/glutenproject/execution/TakeOrderedAndProjectExecTransform.scala [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    } else {
+      val codegenStageCounter: AtomicInteger = ColumnarCollapseCodegenStages.codegenStageCounter
+      val sortExecPlan = SortExecTransformer(sortOrder, false, new ColumnarInputAdapter(child))
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/execution/TakeOrderedAndProjectExecTransform.scala:72]: `child` already is a `WholeStageTransformerExec`, can remove this `WholeStageTransformerExec` and then add a new one for sort, limit ?

**Comment by lgbo-ustc** [comment]
@rui-mo please check what is wrong in compiling velox
```
-- Found folly: /usr/local
CMake Error at velox/CMakeLists.txt:58 (message):
  Velox library not exists:
  /opt/gluten/ep/build-velox/build/velox_ep/_build/release/velox/vector/arrow/c/libvelox_arrow_bridge.a
Call Stack (most recent call first):
```

**Comment by rui-mo** [comment]
> @rui-mo please check what is wrong in compiling velox
> 
> ```
> -- Found folly: /usr/local
> CMake Error at velox/CMakeLists.txt:58 (message):
>   Velox library not exists:
>   /opt/gluten/ep/build-velox/build/velox_ep/_build/release/velox/vector/arrow/c/libvelox_arrow_bridge.a
> Call Stack (most recent call first):
> ```

Velox backend needs this PR https://github.com/oap-project/gluten/pull/637, which will be merged soon. Maybe you can rebase after that, thanks.

**Comment by lgbo-ustc** [comment]
@rui-mo I don't find the failed ut in the velox logs. Could you check it?

**Review by zzcclp — APPROVED** [review]
LGTM

**Pull request merged**